agents/utils/helpers.py

```diff
@@ -0,0 +1,87 @@
+import logging
+import re
+import subprocess
+import time
+
+from .logtools import maybe_log_message
+
+
+def restart_service(logger, fallback_logger, service, attempts=3):
+    """
+    Attempts to restart a system service with exponential backoff
+    if it is found to be inactive. Logs each attempt and result.
+
+    Parameters:
+        agent (ServerAgent): Agent instance used for logging. Must have
+        `logger` and `fallback_logger` attributes.
+        service (str): Name of the system service to restart (e.g., 'ssh').
+
+    Returns:
+        bool: True if the service was restarted, False otherwise.
+
+    Notes:
+        This function uses 'sudo systemctl restart <service>' and expects
+        that the agent has sufficient privileges to perform the operation.
+        Backoff strategy uses 2s, 4s, and 8s delays between attempts.
+    """
+
+    maybe_log_message(
+        '%s not active. Attempting restart...' % service,
+        logger,
+        fallback_logger=fallback_logger
+        )
+
+    for i in range(1, attempts+1):
+        try:
+            delay = 2**i
+
+            maybe_log_message(
+                'Restarting %s (delay before restart: %s).' % (service, delay),
+                logger,
+                fallback_logger=fallback_logger
+                )
+
+            time.sleep(delay)
+
+            retcode = subprocess.call([
+                'sudo',
+                'systemctl',
+                'restart',
+                service
+                ])
+
+            if retcode == 0:
+                maybe_log_message(
+                    '%s service restarted successfully.' % service,
+                    logger,
+                    fallback_logger=fallback_logger,
+                    level=logging.INFO
+                    )
+                return True  # If restsrting successful
+            else:
+                maybe_log_message(
+                    '%s restart failed with code %s.' % (service, retcode),
+                    logger,
+                    fallback_logger=fallback_logger
+                    )
+
+        except Exception as restart_err:
+            maybe_log_message(
+                'Error during %s service restart: %s' % (
+                    service,
+                    restart_err),
+                logger,
+                fallback_logger=fallback_logger,
+                exc_info=True
+                )
+            return False  # Stop after first fatal error
+    return False  # If restarting failed
+
+
+def is_valid_ip(output):
+    """
+    Validate if the output is a correctly formatted IPv4 address.
+    Returns:
+        bool: True if the output is a valid IP address, False otherwise.
+    """
+    return re.match(r'^\d{1,3}(\.\d{1,3}){3}$', output.strip()) is not None
```

Comment:
Your IP regex works for basic patterns like 192.168.0.1 but also accepts invalid ones like 999.999.999.999.
Consider using `socket.inet_aton()` for more accurate validation:
```
import socket

def is_valid_ip(output):
    try:
        socket.inet_aton(output.strip())
        return True
    except socket.error:
        return False
```
Or if you follow `regex` leveraging, refine it to:
```
r'^(25[0-5]|2[0-4][0-9]|[01]?[0-9]?[0-9])(\.(?!$)){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9]?[0-9])$'
```